### ISSUE TITLE
Downlink manual frame

### ISSUE THREAD

**Issue by jot4p** [gotthardp/lorawan-server#393]
Hi, 

It's very common in new sensors send directly one downlink frame in order to configure the sensors (like https://www.nasys.no/ sensors) 

This is necessery only one time, and I need to create a simple php callback server in order to create my downlink response/configuration.

It's possible to add new feature to server to create a simple downlink frame for sensor ? or this exists? 

I need to send a simple frame in HEX for port 50 or other .. like this (Pag 7):
https://www.nasys.no/wp-content/uploads/Pulse-Analog-Reader_UM3023_1.pdf

Thanks.

**Comment by gotthardp** [comment]
Hi. The uplink that triggers this config-downlink always the same? Is it feature to "always respond to THIS uplink with THAT downlink" or to "as a response to the NEXT uplink (whatever it will be) send THIS downlink"?

**Comment by jot4p** [comment]
Imagine this..  I have 1 lorawan datalogger for measure water consumption.. after first connection I need to send my settings, number of times to sync per day, number of pulses to read, read in liters  or m3, this is sended with a simple frame like: "001111001100110011001100111111111110000" (example only) or the same in HEX for a specific port (like 50 or other)
I need to send this simple command to the sensor...   maybe i need to send this at NEXT uplink

**Comment by gotthardp** [comment]
How about to extend the Parse Uplink handler so it can not only forward data to a connector, but also to respond-- send immediately a downlink back? You would need to define (in Erlang) your own logic about when the downlink should be sent, or just send it always when a certain uplink is sent.

**Comment by jot4p** [comment]
Hi, 
Yes, second part..  is the response for any frame only 1 time (after that sensor is configured) and I don't need to config again.

**Comment by jot4p** [comment]
Ok.. I need to send this : 01A8C0830000000000000000 for the next time that sensor connects to lora server.. it's possible ?  Please give me an example. 
Thanks

**Comment by gotthardp** [comment]
Sending something upon next uplink is a standard downlink feature. Yes, the image shows the frame to be sent. It was created by sending a downlink request to a connector. See https://github.com/gotthardp/lorawan-server/blob/master/doc/Handlers.md#downlink

**Comment by jot4p** [comment]
Ok Perfect. 

It's the same send in HEX or BIN for the sensor ??

**Comment by gotthardp** [comment]
you always enter binary data as a HEX string, unless the binary connector is used

**Comment by hallard** [comment]
@jot4p 
I do this with sample code based on websocket example `ws.html` from the server
[basic sample](https://github.com/gotthardp/lorawan-server/blob/master/priv/admin/ws.html)
![image](https://user-images.githubusercontent.com/2471931/40588143-59ff0ed2-61d9-11e8-9ff1-d362aada5fe8.png)

**Comment by jot4p** [comment]
Thanks. Perfect.. 
I use Insomnia to create the post and it's solve my problem until next verson of your server.

**Comment by jot4p** [comment]
Hi,

My downlink is processed
![image](https://user-images.githubusercontent.com/15892047/41040911-9fb548ce-6995-11e8-85c5-3a0bbf200d77.png)

The downlink is sent but the sensor does not respond to my downlink.
The manufacturer confirms that it is correct (HEX):
> 00070F003C0001
>  010407028C

What is wrong ?

**Comment by gotthardp** [comment]
Can you see from the device log it was received by the device?

**Comment by jot4p** [comment]
No, my device is an "NAS UM3023A" 
https://www.nasys.no/wp-content/uploads/UM3023-FW-ver-0.5.0.pdf

If Isend the correct frame the device return a response.

**Comment by hallard** [comment]
@jot4p I don't see any port on your downlink

![image](https://user-images.githubusercontent.com/2471931/41044128-cdd7c762-69a5-11e8-955b-8b9850fe44d5.png)

No port, no data processing ;-)
Should be port 50 configuration message for your device

**Comment by jot4p** [comment]
![image](https://user-images.githubusercontent.com/15892047/41046977-6c59ee2c-69a3-11e8-8059-a7c2e62e11c4.png)

I have the port now :) (old example) 

but not work

**Comment by hallard** [comment]
Looks good,
may we have the frame received screenshot when the downlink is send?

**Comment by hallard** [comment]
I mean this screen for Web Interface (refreshed with communications between the server and your node)
![image](https://user-images.githubusercontent.com/2471931/41051774-52384426-69b7-11e8-9aef-f9695a5107fe.png)

**Comment by jot4p** [comment]
![image](https://user-images.githubusercontent.com/15892047/41051857-2ec3e4ee-69af-11e8-98ef-4bfe1e5d41a3.png)

**Comment by hallard** [comment]
ok thanks, your device joined and sending port 99 is the starting frame of nasys devices
now the same with seeing downlink send?

**Comment by jot4p** [comment]
Yes... the server send the packet and after that the sensor do not talk anymore. 

The frame "00" for port 49 make the sensor return "device frame" but I send and the sensor not return anything.

**Comment by gotthardp** [comment]
If you sent a downlink (that was received by the device) and then the device stopped talking, then the server sent some ADR config data that misaligned the device configuration. Check the device profile configuration.

**Comment by jot4p** [comment]
Hi, 

The creator of the sensor confirm:

´´´
Frame 1:
{
	"data": "010407020C",
	"deveui": "70B3D5B020037081",
	"port": "50"
}

Frame 2:
{
	"data": "00070F003C0001",
	"deveui": "70B3D5B020037081",
	"port": "50"
}

Confirm Frame:
{
	"data": "00",
	"deveui": "70B3D5B020037081",
	"port": "49"
} ´´´ 

![image](https://user-images.githubusercontent.com/15892047/41902432-524a8dc4-792b-11e8-8459-8dd53e89a3c3.png)


After the manual procedure of sending status data (1 second with magnet), the downlink's disappear from the server (are processed and sended to the sensor ??) 

![image](https://user-images.githubusercontent.com/15892047/41902519-8c470228-792b-11e8-967c-6297f135da24.png)

This is the response from the node correct  ? 

The creator of the sensor replied:

"It looks like the message is not received by the device. 
The second message is retransmit of the status message, as the ack is not received. If i’m correct then one f the values is fcnt and it is same for both packets. So there is no reply to fport49 message. (it should also be sent on fPort49). Is there a change you could use RX2 window instead of RX1 to send a message to the node. (in some networks it is possible to force it to RX2).  And then try again.
"

What do you think ? 

Thanks.

**Comment by gotthardp** [comment]
Yes, the received frames are retransmissions. Yes, you can force RX2 in my server too (see the RX Window in the Profile setting).

**Comment by jot4p** [comment]
but what is the problem ? the sensor receive the correct frame  ? How I can confirm that ?

**Comment by gotthardp** [comment]
Ideally you should get debug console of the device and have a look if the downlink was received and if it was accepted by the device. Debugging devices without any logs is always hard and frustrating.

**Comment by jot4p** [comment]
the default settings of the sensor:

```
ADR is supported. ADR can change the DR from SF12 to SF7. And TX power from 0-7.

The LoRaWAN setting are according to the LoRaWAN regional specs.

RECEIVE_DELAY1 1 s 
RECEIVE_DELAY2 2 s (must be RECEIVE_DELAY1 + 1s) 
JOIN_ACCEPT_DELAY1 5 s 
JOIN_ACCEPT_DELAY2 6 s 
MAX_FCNT_GAP 16384 
ADR_ACK_LIMIT 64 
ADR_ACK_DELAY 32 
ACK_TIMEOUT 2 +/- 1 s (random delay between 1 and 3 seconds) 

```

**Comment by gotthardp** [comment]
What server version are you using?
As a safety measure I'd recommend to set "Set Rx2 freq" field to the default value of your Network. There is a bug (I didn't find yet) that may cause some issues.

**Comment by jot4p** [comment]
My version is 0.53 
I dont set Rx2 value this is automatic ...  you can see here:
https://user-images.githubusercontent.com/15892047/41917685-e0703622-7952-11e8-869f-8e91e3709965.png

**Comment by gotthardp** [comment]
I recommend you set this value too.

**Comment by jot4p** [comment]
I remove my device and create new entry on commisioned list:

![image](https://user-images.githubusercontent.com/15892047/42167927-65ea400a-7e07-11e8-8ad4-b3c84cdcda14.png)

After that (configured with manufacture settings) my device try to join but is not possible !! 
why ?  
![image](https://user-images.githubusercontent.com/15892047/42167961-8a9b86c0-7e07-11e8-9ac6-c4ce7797c63f.png)

![server_error](https://user-images.githubusercontent.com/15892047/42167886-4b99174e-7e07-11e8-94d5-754b7c67e2bf.png)

**Comment by jot4p** [comment]
my config in multitech packet forwarder
```

{
	"SX1301_conf": {
		"antenna_gain": 8,
		"chan_FSK": {
			"bandwidth": 125000,
			"datarate": 50000,
			"enable": true,
			"if": 300000,
			"radio": 1
		},
		"chan_Lora_std": {
			"bandwidth": 250000,
			"enable": true,
			"if": -200000,
			"radio": 1,
			"spread_factor": 7
		},
		"chan_multiSF_0": {
			"enable": true,
			"if": -400000,
			"radio": 1
		},
		"chan_multiSF_1": {
			"enable": true,
			"if": -200000,
			"radio": 1
		},
		"chan_multiSF_2": {
			"enable": true,
			"if": 0,
			"radio": 1
		},
		"chan_multiSF_3": {
			"enable": true,
			"if": -400000,
			"radio": 0
		},
		"chan_multiSF_4": {
			"enable": true,
			"if": -200000,
			"radio": 0
		},
		"chan_multiSF_5": {
			"enable": true,
			"if": 0,
			"radio": 0
		},
		"chan_multiSF_6": {
			"enable": true,
			"if": 200000,
			"radio": 0
		},
		"chan_multiSF_7": {
			"enable": true,
			"if": 400000,
			"radio": 0
		},
		"clksrc": 0,
		"lbt_cfg": {
			"enable": false,
			"rssi_target": 160
		},
		"lorawan_public": true,
		"radio_0": {
			"enable": true,
			"freq": 867500000,
			"rssi_offset": -162,
			"tx_enable": true,
			"tx_freq_max": 870000000,
			"tx_freq_min": 863000000,
			"type": "SX1257"
		},
		"radio_1": {
			"enable": true,
			"freq": 868500000,
			"rssi_offset": -162,
			"tx_enable": true,
			"type": "SX1257"
		},
		"tx_lut_0": {
			"dig_gain": 0,
			"mix_gain": 11,
			"pa_gain": 0,
			"rf_power": -6
		},
		"tx_lut_1": {
			"dig_gain": 0,
			"mix_gain": 13,
			"pa_gain": 0,
			"rf_power": -3
		},
		"tx_lut_10": {
			"dig_gain": 0,
			"mix_gain": 15,
			"pa_gain": 2,
			"rf_power": 16
		},
		"tx_lut_11": {
			"dig_gain": 0,
			"mix_gain": 10,
			"pa_gain": 3,
			"rf_power": 20
		},
		"tx_lut_12": {
			"dig_gain": 0,
			"mix_gain": 12,
			"pa_gain": 3,
			"rf_power": 23
		},
		"tx_lut_13": {
			"dig_gain": 0,
			"mix_gain": 13,
			"pa_gain": 3,
			"rf_power": 25
		},
		"tx_lut_14": {
			"dig_gain": 0,
			"mix_gain": 15,
			"pa_gain": 3,
			"rf_power": 26
		},
		"tx_lut_15": {
			"dig_gain": 0,
			"mix_gain": 15,
			"pa_gain": 3,
			"rf_power": 27
		},
		"tx_lut_2": {
			"dig_gain": 0,
			"mix_gain": 9,
			"pa_gain": 1,
			"rf_power": 0
		},
		"tx_lut_3": {
			"dig_gain": 0,
			"mix_gain": 10,
			"pa_gain": 1,
			"rf_power": 3
		},
		"tx_lut_4": {
			"dig_gain": 0,
			"mix_gain": 12,
			"pa_gain": 1,
			"rf_power": 6
		},
		"tx_lut_5": {
			"dig_gain": 0,
			"mix_gain": 10,
			"pa_gain": 2,
			"rf_power": 10
		},
		"tx_lut_6": {
			"dig_gain": 0,
			"mix_gain": 11,
			"pa_gain": 2,
			"rf_power": 11
		},
		"tx_lut_7": {
			"dig_gain": 0,
			"mix_gain": 11,
			"pa_gain": 2,
			"rf_power": 12
		},
		"tx_lut_8": {
			"dig_gain": 2,
			"mix_gain": 12,
			"pa_gain": 2,
			"rf_power": 13
		},
		"tx_lut_9": {
			"dig_gain": 0,
			"mix_gain": 13,
			"pa_gain": 2,
			"rf_power": 14
		}
	},
	"gateway_conf": {
		"forward_crc_disabled": false,
		"forward_crc_error": false,
		"forward_crc_valid": true,
		"gateway_ID": "00800000A00005C9",
		"keepalive_interval": 10,
		"push_timeout_ms": 100,
		"serv_port_down": 1680,
		"serv_port_up": 1680,
		"server_address": "loraserver.dstinovacao.com",
		"stat_interval": 30,
                 "autoquit_threshold": 60
	}
}
```
